quinn/src/connection.rs

```diff
@@ -70,29 +70,55 @@ impl Connecting {
 
     /// Convert into a 0-RTT or 0.5-RTT connection at the cost of weakened security
     ///
-    /// Opens up the connection for use before the handshake finishes, allowing the API user to
-    /// send data with 0-RTT encryption if the necessary key material is available. This is useful
-    /// for reducing start-up latency by beginning transmission of application data without waiting
-    /// for the handshake's cryptographic security guarantees to be established.
+    /// Returns `Ok` immediately if the local endpoint is able to attempt 0/0.5-RTT conversion. If
+    /// so, the returned [`Connection`] can be used to open streams and write data to them without
+    /// waiting for the rest of the handshake to complete, at the cost of weakened cryptographic
+    /// security guarantees. The returned [`ZeroRttAccepted`] future resolves when the handshake
+    /// does complete, at which point subsequently opened streams and written data will have full
+    /// cryptographic protection.
     ///
-    /// When the `ZeroRttAccepted` future completes, the connection has been fully established.
+    /// Outgoing
+    /// ---
     ///
-    /// # Security
+    /// For outgoing connections, the initial attempt to convert to 0-RTT will proceed if the
+    /// [`crypto::ClientConfig`][crate::crypto::ClientConfig] attempts to resume a previous TLS
+    /// session. However, **the remote endpoint may not actually _accept_ the 0-RTT data**--yet
+    /// still accept the connection attempt in general. This may occur, for example, if the remote
+    /// endpoint has restarted since the local endpoint last connected to it.
     ///
-    /// On outgoing connections, this enables transmission of 0-RTT data, which might be vulnerable
-    /// to replay attacks, and should therefore never invoke non-idempotent operations.
+    /// This possibility is conveyed through the [`ZeroRttAccepted`] future--when the handshake
+    /// completes, it resolves to true if the 0-RTT data was accepted and false if it was rejected.
+    /// If it was rejected, streams opened prior to the handshake completing will _hang_, and their
+    /// existence will typically not be conveyed to the remote application.
     ///
-    /// On incoming connections, this enables transmission of 0.5-RTT data, which might be
-    /// intercepted by a man-in-the-middle. If this occurs, the handshake will not complete
-    /// successfully.
+    /// By default, quinn uses [`rustls::ClientConfig`] as the client crypto implementation. It
+    /// allows the initial attempt to convert to 0-RTT to proceed if its `enable_early_data` field
+    /// is set to true and its `resumption` field recognizes the server name (which defaults to an
+    /// in-memory cache of 256 server names). **If you are manually providing the client crypto
+    /// config, you must set `enable_early_data` to true to get 0-RTT working.**
```

Comment:
I don't think this should be documented here.

Reply:
Moved.